File: app/src/main/java/com/softwareverde/tidyduck/Review.java
package com.softwareverde.tidyduck;

import com.softwareverde.tidyduck.most.FunctionBlock;
import com.softwareverde.tidyduck.most.FunctionCatalog;
import com.softwareverde.tidyduck.most.MostFunction;
import com.softwareverde.tidyduck.most.MostInterface;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class Review {
    private Long _id;
    private FunctionCatalog _functionCatalog = null;
    private FunctionBlock _functionBlock = null;
    private MostInterface _mostInterface = null;
    private MostFunction _mostFunction = null;
    private Account _account;
    private String ticketUrl;
    private Date _createdDate;
    private List<ReviewVote> _reviewVotes = new ArrayList<>();
    private List<ReviewComment> _reviewComments = new ArrayList<>();

    public Long getId() {
        return _id;
    }

    public void setId(final Long id) {
        _id = id;
    }

    public FunctionCatalog getFunctionCatalog() {
        return _functionCatalog;
    }

    public void setFunctionCatalog(final FunctionCatalog functionCatalog) {
        _functionCatalog = functionCatalog;
    }

    public FunctionBlock getFunctionBlock() {
        return _functionBlock;
    }

    public void setFunctionBlock(final FunctionBlock functionBlock) {
        _functionBlock = functionBlock;
    }

    public MostInterface getMostInterface() {
        return _mostInterface;
    }

    public void setMostInterface(final MostInterface mostInterface) {
        _mostInterface = mostInterface;
    }

    public MostFunction getMostFunction() {
        return _mostFunction;
    }

    public void setMostFunction(final MostFunction mostFunction) {
        _mostFunction = mostFunction;
    }

    public Account getAccount() {
        return _account;
    }

    public void setAccount(final Account account) {
        _account = account;
    }

    public String getTicketUrl() {
        return ticketUrl;
    }

    public void setTicketUrl(final String ticketUrl) {
        this.ticketUrl = ticketUrl;
    }

    public Date getCreatedDate() {
        return _createdDate;
    }

    public void setCreatedDate(final Date createdDate) {
        _createdDate = createdDate;
    }

    public List<ReviewVote> getReviewVotes() {
        return new ArrayList<>(_reviewVotes);
    }

    public void addReviewVote(final ReviewVote reviewVote) {
        _reviewVotes.add(reviewVote);
    }

    public void setReviewVotes(final List<ReviewVote> reviewVotes) {
        _reviewVotes = new ArrayList<>(reviewVotes);
    }

    public List<ReviewComment> getReviewComments() {
        return new ArrayList<>(_reviewComments);
    }

    public void addReviewComment(final ReviewComment reviewComment) {
        _reviewComments.add(reviewComment);
    }

    public void setReviewComments(final List<ReviewComment> reviewComments) {
        _reviewComments = new ArrayList<>(reviewComments);
    }
}

File: app/src/main/webapp/js/classes/review.js
class Review {
    static fromJson(json) {
        let functionCatalog = null;
        let functionBlock = null;
        let mostInterface = null;
        let mostFunction = null;

        if (json.functionCatalogId) {
            functionCatalog = new FunctionCatalog();
            functionCatalog.setId(json.functionCatalogId);
        }
        if (json.functionBlockId) {
            functionBlock = new FunctionBlock();
            functionBlock.setId(json.functionBlockId);
        }
        if (json.mostInterfaceId) {
            mostInterface = new MostInterface();
            mostInterface.setId(json.mostInterfaceId);
        }
        if (json.mostFunctionId) {
            mostFunction = new MostFunction();
            mostFunction.setId(json.mostFunctionId);
        }

        const account = new Account();
        account.setId(json.accountId);

        const ticketUrl = json.ticketUrl;

        const reviewVotes = [];
        const reviewVotesJson = json.reviewVotes;
        for (let i in reviewVotesJson) {
            const reviewVoteJson = reviewVotesJson[i];
            const reviewVote = ReviewVote.fromJson(reviewVoteJson);
            reviewVotes.push(reviewVote);
        }

        const reviewComments = [];
        const reviewCommentsJson = json.reviewComments;
        for (let i in reviewCommentsJson) {
            const reviewCommentJson = reviewCommentsJson[i];
            const reviewComment = ReviewComment.fromJson(reviewCommentJson);
            reviewComments.push(reviewComment);
        }

        const review = new Review();

        review.setId(json.id);
        review.setFunctionCatalog(functionCatalog);
        review.setFunctionBlock(functionBlock);
        review.setMostInterface(mostInterface);
        review.setMostFunction(mostFunction);
        review.setCreatedDate(json.createdDate);
        review.setAccount(account);
        review.setTicketUrl(ticketUrl);
        review.setReviewVotes(reviewVotes);
        review.setReviewComments(reviewComments);

        return review;
    }

    static toJson(review) {
        const id = review.getId();
        const functionCatalogId = review.getFunctionCatalog() ? review.getFunctionCatalog().getId() : null;
        const functionBlockId = review.getFunctionBlock() ? review.getFunctionBlock().getId() : null;
        const mostInterfaceId = review.getMostInterface() ? review.getMostInterface().getId() : null;
        const mostFunctionId = review.getMostFunction() ? review.getMostFunction().getId() : null;
        const accountId = review.getAccount().getId();
        const ticketUrl = review.getTicketUrl();
        const createdDate = review.getCreatedDate();

        return {
            id:                 id,
            functionCatalogId:  functionCatalogId,
            functionBlockId:    functionBlockId,
            mostInterfaceId:    mostInterfaceId,
            mostFunctionId:     mostFunctionId,
            accountId:          accountId,
            ticketUrl:          ticketUrl,
            createdDate:        createdDate
        };
    }

    constructor() {
        this._id = null;
        this._functionCatalog = null;
        this._functionBlock = null;
        this._mostInterface = null;
        this._mostFunction = null;
        this._account = null;
        this._ticketUrl = null;
        this._createdDate = null;
        this._reviewVotes = [];
        this._reviewComments = [];
    }

    getReviewObject() {
        return this._functionCatalog || this._functionBlock || this._mostInterface || this._mostFunction;
    }

    getReviewName() {
        const reviewObject = this.getReviewObject();
        return reviewObject.getName();
    }

    setId(id) {
        this._id = id;
    }

    getId() {
        return this._id;
    }

    setFunctionCatalog(functionCatalog) {
        this._functionCatalog = functionCatalog;
    }

    getFunctionCatalog() {
        return this._functionCatalog;
    }

    setFunctionBlock(functionBlock) {
        this._functionBlock = functionBlock;
    }

    getFunctionBlock() {
        return this._functionBlock;
    }

    setMostInterface(mostInterface) {
        this._mostInterface = mostInterface;
    }

    getMostInterface() {
        return this._mostInterface;
    }

    setMostFunction(mostFunction) {
        this._mostFunction = mostFunction;
    }

    getMostFunction() {
        return this._mostFunction;
    }

    setAccount(account) {
        this._account = account;
    }

    getAccount() {
        return this._account;
    }

    setTicketUrl(ticketUrl) {
        this._ticketUrl = ticketUrl;
    }

    getTicketUrl() {
        return this._ticketUrl;
    }

    setCreatedDate(createdDate) {
        this._createdDate = createdDate;
    }

    getCreatedDate() {
        return this._createdDate;
    }

    setReviewVotes(reviewVotes) {
        this._reviewVotes = reviewVotes;
    }

    addReviewVote(reviewVote) {
        this._reviewVotes.push(reviewVote);
    }

    getReviewVotes() {
        return this._reviewVotes;
    }

    setReviewComments(reviewComments) {
        this._reviewComments = reviewComments;
    }

    addReviewComment(reviewComment) {
        this._reviewComments.push(reviewComment);
    }

    getReviewComments() {
        return this._reviewComments;
    }
}